### ISSUE TITLE
Simplify onboarding by moving/deferring location questions to relevant Tasks (for Poppy, Dakota)

### ISSUE THREAD

**Issue by hinzed1127** [hinzed1127/pivotal-to-gh-importer#275]
This issue was imported from Pivotal Tracker.

  | Field | Value |
  | --- | --- |
  | Story ID | [#183444249](https://www.pivotaltracker.com/story/show/183444249) |
  | State | accepted |
  | Requested By | dillonsullivan |
  | Owned By | Anne LoVerso |
  | Estimate | 4 |
  | Priority | p3 - Low |
  | Created At | Oct 3, 2022 |
  | Accepted At | Nov 16, 2022 |
      
## User Stories

...

## Background

Questions only appear to top of tasks for non-SP/GP in the guest mode and needs to form phase:

- Removing location question from onboarding (eliminate entire step, subsequent steps will need to be updated)
- Adding location question to top of the 4 impacted tasks: Get your Mercantile License (if town requires it), Confirm Liquor License Availability, Get Your Liquor and Mercantile License, County Clerk Register Trademark (SP/GP) (where appropriate)
   - For the tasks they need to determine wether to show the location question
    - For non-SP/GP, these questions appear 1) in Guest Mode Needs to Form and Needs to Form phases only, because the location will have been collected by later phases, 2) if location has not been entered yet by the user in any place (e.g., previously entering it in one of these Tasks, or the Profile screen, or by forming via API or formation modal)
    - For SP/GP, these questions appear 1) in Guest Mode and Needs to Register Phase, as long as the user hasn't answered it yet in a Task or on the Profile screen, or in the Register for Taxes Modal
  - Anne note: the above bullets are a long way of saying "given the location question has not yet been answered, then it appears in these 4 tasks" and "given the location question has been answered (at any point in the app) then the question does not appear" 
- Add the location question after the business effective date question in the manual formation modal if necessary (for non-SP/GP who haven't already answered)
    - This question only appears in the formation modal if the Location has not been answered yet by the user at all in any fashion
- Add the location question to the manual Register for Taxes modal if necessary
    - This question only appears in the Register for Taxes modal if the Location has not been answered yet by the user at all in any fashion
- If location questions are added, what if any updates are needed to the user profile??
    - User profile must now consider the situation where the user has not entered their Location at all yet, and should allow the user to enter their Location for the first time in the Profile if they choose
- Formation form in the formation task should also handle when/if the location has been provided
    - If user has not answered the Location question anywhere when attempting formation via the API, formation form in the formation Task must make sure to allow the user to specify their Location for the first time (instead of it automatically being pre-populated). If the user is setting it for the first time via Formation, this should similarly answer the question in the Profile and elsewhere in Navigator.

## Additional Notes

- Can you change the address in your profile? Do we treat it as required? Once added, you can update it, but shouldn't allow a blank location. Once saved, treat as a required field. 
- The formation form... if the user has not already provided location, the form must require them to supply the location and store in the profile.
- Not required upfront, but required to graduate.

## Designs

https://www.figma.com/file/IKcrTIyvaS593LuJbAVFR3/39---43.-Navigator-Designs?node-id=18857%3A381657

## Analytics

Add to all (4) Tasks, when location question appears and is successfully submitted

`task_location_question/submit/location_entered_for_first_time`

Add to Profile, when and only when the user has entered and submitted a value into the Location question for the first time

`profile_location_question/submit/location_entered_for_first_time`

Add to formation form, if the user had not previously specified Location elsewhere in Navigator but is now specifying it via successfully submission of the Formation API form

`business_formation_location_question/submit/location_entered_for_first_time`

## Comments
@fiorellagalvez can you please add county clerk designs (dillonsullivan - Oct 11, 2022)

@dillonsullivan I'm blocked on that, I need a link to the actual task. I'm connecting with @mikewolfd to get it since he identified all the location-impacted tasks. (Fiorella Galvez - Oct 12, 2022)

@fiorellagalvez @caryeuwer @amandabeall I have a design question here, I see this has a link in the content to "Edit this city in your profile" - is this link desired if there is an "Edit City" link directly above? Why would we want to send the user all the way to the profile page if they could just click the other one to edit inline here on the same page? 
 (Anne LoVerso - Oct 18, 2022)

@anneloverso_nj I agree, we should not have both links. If the success alert does not disappear, we don't need the "Edit this city in your profile link". But if it does disappear, we do want to inform the user they can edit their city in the profile. cc/ @caryeuwer (Fiorella Galvez - Oct 18, 2022)

@fiorellagalvez @anneloverso_nj that's right, it's only necessary if the success message goes away. Won't the success message go away if the user leaves the task and returns? It would be weird if they had an old success message (caryeuwer - Oct 19, 2022)

@caryeuwer that's part of the questions I have. for other tasks with input fields (EIN, naics code, etc) we keep the success message on the task screen permanently (Anne LoVerso - Oct 19, 2022)

Notes from 16th minute:
- when page is reloaded or revisited, the green "success" banner goes away but the content remains
- we're removing the "Edit this in your profile" link altogether (Anne LoVerso - Oct 19, 2022)

@ej3 @dillonsullivan Currently, the logic for Dakota Nexus is that we only ask them the Location question in onboarding if they answer "yes" to "will you own/lease space in NJ"

How will this logic impact deferring the location question for dakota?  (Anne LoVerso - Oct 19, 2022)

@caryeuwer @amandabeall @ceciliaramirez1 @jessica2I I added reviews for Product, Content, and Design for this -please let me know if that's not appropriate. Otherwise, do you all want to review this together? (for my benefit at least) 

Cc @ej3 @anneloverso_nj @joshuaalley  (dillonsullivan - Oct 25, 2022)

@dillonsullivan I don't think this is ready for review yet? @joshuaalley 
 (Anne LoVerso - Oct 25, 2022)

@dillonsullivan @ej3 @caryeuwer @fiorellagalvez can we further flesh out the AC for what changes need to be made to the profile? Specifically, at what point does location become a required field in the profile? I assume it's dependent on phase, right, in some phases you can remove it and in some you cannot. And separately, is it ever locked or is it always allowed to be changed? (joshuaalley - Oct 27, 2022)

Currently, the assumption is that after you've completed onboarding, you have a Location. Thus, the Profile currently lets you change your Location, but not delete your location. That's the current state of affairs.

After this story is implemented, you'll be entered from onboarding into "Needs to Form" Phase without a location as an LLC/Corp/LLP/LP, or into "Needs to Register" Phase without a location as an SP/GP. In graduating from either of these Phases from the next one, you'll be required to enter your location, and thus from that next Phase forward, it should behave as it does now: you can change it in the Profile, but not delete it.

The open question is how the Profile should behave while you're in the respective Phase for your legal structure when the location is still "optional." @caryeuwer @fiorellagalvez @jessica2I @dillonsullivan would throw this question to you… if you're still in "Needs to Form" as an LLC, for example, should you be able to add your Location then delete it again? Or should we not allow you to delete it once it's added, even for these Phases that can exist "without" it initially? (ej3 - Oct 27, 2022)

@ej3 @caryeuwer @fiorellagalvez just to add to that, according to the designs we've been building against, there does exist a "remove" button inside of the tasks which implies it does start out deleteable? (Anne LoVerso - Oct 28, 2022)

@ej3 i don't think one should be allowed to delete it once added, only change it.

@anneloverso_nj I suggest then that we eliminate the "Remove" button for deferred location tasks (i can update the designs accordingly)

@joshuaalley location should be locked if you enter the location via formation api. A tooltip explaining why that location is locked should be displayed to the right of the field label, similar to a locked business effective date (lmk if you need design for this)
 (caryeuwer - Oct 28, 2022)

@caryeuwer Sounds like the simplest/cleanest approach. (ej3 - Oct 28, 2022)

Not sure if @jessica2I has an alternative thought on removing location after adding it, but assuming she doesn't, does this answer what you need? @anneloverso_nj @joshuaalley  (ej3 - Oct 28, 2022)

Can we standardize this, so we ask them both in the same phase - need to register phase for example? 
In terms of offer option to edit/delete - it should follow the same standardization we have for other things. Don't we let them/plan to let them do both for date of formation, name etc. 
Lastly, can we lock the location if they formed through us?  Or need that be a separate tkt. (jessica2I - Oct 28, 2022)

Per 16th minute discussion with @jessica2I @caryeuwer @ej3 :

For the phases where your location is optional and you might not have entered it yet: you can remove the location using either the "Remove" button on the task or the profile. 

At the phase at which the location becomes required, then the profile should throw an error and not allow the location to be removed - it becomes a required field (Anne LoVerso - Oct 31, 2022)

@caryeuwer @fiorellagalvez Where in the Tax Registration Modal should the Location question appear (for when it does appear?)

Right now the fields in order are:
- Ownership types
- Number of employees

Should it be before? Middle? After? thanks (Anne LoVerso - Nov 2, 2022)

@anneloverso_nj I think just adding after # of employees should be fine. Unless @jessica2I would like it in a particular order. (Fiorella Galvez - Nov 2, 2022)

That's fine.  (jessica2I - Nov 2, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/c4efeb2c270edc76b8d72be327aad5994fa9febf
navigator.business.nj.gov

feat: [#183444249] defer onboarding location question (Anne LoVerso - Nov 10, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/b7544169b1b09b66a835f76a6d19cd7c62d89e6a
navigator.business.nj.gov

feat: [#183444249] defer onboarding location question (Anne LoVerso - Nov 14, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/1d4fa94545a7e3b6dfb30721bf9b7074c6747c0a
navigator.business.nj.gov

feat: [#183444249] defer onboarding location question (Anne LoVerso - Nov 14, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/7d66b8b3451030799bb98b61611a8cde2ee0c267
navigator.business.nj.gov

feat: [#183444249] defer onboarding location question (Anne LoVerso - Nov 14, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/f2948d60c6d1c1d9a373ee1ec31b2ddeb3788437
navigator.business.nj.gov

feat: [#183444249] defer onboarding location question (Anne LoVerso - Nov 14, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/3fc4e3f25297d88805aa8757b633d1a18b64c378
navigator.business.nj.gov

feat: [#183444249] defer onboarding location question (Anne LoVerso - Nov 14, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/78b3c5475a34892dc6dad5dd76b0704d8c31cf71
navigator.business.nj.gov

feat: [#183444249] defer onboarding location question (Anne LoVerso - Nov 14, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/1e174dd3f6b0e3d43e541ff2d35f16f825fff674
navigator.business.nj.gov

feat: [#183444249] defer onboarding location question (Anne LoVerso - Nov 14, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/1cf2ffda4cdcd2ba4cbd9f5455212889aea7d104
navigator.business.nj.gov

feat: [#183444249] defer onboarding location question (Anne LoVerso - Nov 14, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/9ba027504f8414a2e24897b1600f6bc6a9a3ad5a
navigator.business.nj.gov

feat: [#183444249] defer onboarding location question (Anne LoVerso - Nov 14, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/5b983a57478f0f6ce9821d095d65e5c91fead7d3
navigator.business.nj.gov

feat: [#183444249] defer onboarding location question (Anne LoVerso - Nov 14, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/7d8fdf3e20c618baae25b025de49ddbfde500eb0
navigator.business.nj.gov

feat: [#183444249] defer onboarding location question (Anne LoVerso - Nov 15, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/ad73c743989545b151ba9efcd1fa7f47a1f5f7f4
navigator.business.nj.gov

feat: [#183444249] defer onboarding location question (Anne LoVerso - Nov 15, 2022)

@dillonsullivan @jessica2I @fiorellagalvez ready for review (Anne LoVerso - Nov 15, 2022)

**Content** review set to **pass**

g2g (ceciliaramirez1 - Nov 15, 2022)

@anneloverso_nj can we please adjust the following:
1) Increase the width of the dropdown to match the designs
2) For mobile (320px), can we reduce the space between the dropdown and button (Fiorella Galvez - Nov 15, 2022)

@fiorellagalvez all of our dropdowns and text fields across the site have a max-width on them to purposefully prevent them from spanning across the width of the entire page. See for example the attached screenshot of the profile page with all dropdowns and text fields at the same max-width:  (Anne LoVerso - Nov 15, 2022)

@anneloverso_nj Got it. In that case, can we move the button closer to the dropdown and reduce the space between them?  (Fiorella Galvez - Nov 15, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/321973061a249cbf447a26bfd0dfc3e73f3cfb7d
navigator.business.nj.gov

chore: [#183444249] smaller spacing between deferred location question and save button (Anne LoVerso - Nov 15, 2022)

@fiorellagalvez fixed the spacing (Anne LoVerso - Nov 15, 2022)

@anneloverso_nj looks good on mobile, thank you! Can we adjust it on desktop, move the button closer to the dropdown and bottom align it to the dropdown. It seems to be floating up. 
 (Fiorella Galvez - Nov 15, 2022)

@fiorellagalvez oh good catch! lol changing the bottom padding changed the alignment on desktop now oops. fixing (Anne LoVerso - Nov 15, 2022)

- Given I have selected a public filing business structure 
and I am in the needs to form phase or in guest mode
and the location question has not already been answered
when accessing the "Get your Mercantile License" (if town requires it) task, 
or the "Confirm Liquor License Availability" task, 
or the "Get Your Liquor and Mercantile License task, 
or the business effective date modal on the "Check Available Names and Form Your Business" task,
or the register for state taxes modal on the "Register for State Taxes and Employer Purposes" task,
or the the user profile,
then the "Select Your City" call to action is presented
and upon submitting the location is saved in the profile
and the location question is not asked during on-boarding
or on any other page if already answered

- Given I have selected a trade name business structure 
and I am in the needs to form phase or in guest mode
and the location question has not already been answered
when accessing the "Register a Trade Name" task,
or the "Register for State Taxes and Employer Purposes" task,
or the "County Clerk Register Trademark" task (SP/GP) (where appropriate),
then the "Select Your City" call to action is presented
and upon submitting the location is saved in the profile
and the location question is not asked during on-boarding
or on any other page if already answered (dillonsullivan - Nov 16, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/606d026c83eb8e91d09b894fb0a4feaf328f2d26
navigator.business.nj.gov

chore: [#183444249] fix styling for deferred onboarding question (Anne LoVerso - Nov 16, 2022)

@fiorellagalvez fixed the button styling, ready for re-review (Anne LoVerso - Nov 16, 2022)

**Design** review set to **pass** (Fiorella Galvez - Nov 16, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/e0f226d9c94a2ace1eb80b24bdde963756e4b124
navigator.business.nj.gov

refactor: [#183444249] consolidate helper selectors in tests (Anne LoVerso - Nov 16, 2022)

@anneloverso_nj how do I get the Confirm Liquor License Availability, Get Your Liquor and Mercantile License, County Clerk Register Trademark (SP/GP) (where appropriate) tasks to show up?
 (dillonsullivan - Nov 16, 2022)

@dillonsullivan to get liquor license tasks you need to be a restaurant and answer "yes" to the liquor essential question

"County Clerk Register Trademark" is actually called "Register Trade Name" (Anne LoVerso - Nov 16, 2022)

(dillonsullivan - Nov 16, 2022)

After forming through the Navigator, error upon attempting to update my location (dillonsullivan - Nov 16, 2022)

**Product** review set to **pass** (dillonsullivan - Nov 16, 2022)

@anneloverso_nj I've completed my testing and marked the product review passed, but with two incomplete tasks I'm getting a warning when trying to accept to work item. How would you like to proceed with the 'after push' tasks pending? (dillonsullivan - Nov 16, 2022)

@dillonsullivan those are dev tasks, the ticket can be accepted even with them pending (Anne LoVerso - Nov 16, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/00f163c62819f54be2127f5f759e351ca30492e7
navigator.business.nj.gov

chore: [#183444249] add more e2e tests for deferred location (Anne LoVerso - Nov 16, 2022)

Commit by Anne LoVerso
https://github.com/newjersey/navigator.business.nj.gov/commit/fc65b22834f7f562147a005b363eaa8614b3b7a1
navigator.business.nj.gov

chore: [#183444249] add more e2e tests for deferred location (Anne LoVerso - Nov 16, 2022)

## Referenced Pull Requests
- https://github.com/newjersey/navigator.business.nj.gov/pull/4703
